### PR TITLE
Allow for empty sms

### PR DESCRIPTION
**Pull request by SCdF**
# Description

Allow for empty SMS messages

medic/medic-webapp#3658

# Review checklist

- [ ] Readable: Concise, well named, follows the [style guide](https://github.com/medic/medic-docs/blob/master/development/style-guide.md), documented if necessary.
- [ ] Documented: Announced in Changes.md in plain English. Configuration and user documentation on [medic-docs](https://github.com/medic/medic-docs/)
- [ ] Tested: Unit and/or e2e where appropriate
- [ ] Internationalised: All user facing text
- [ ] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in Changes.md.
- [ ] Webapp CI runs clean against this branch

### CODE REVIEW

File: controllers/record-utils.js [hunk 1]
```diff
@@ -1,7 +1,7 @@
 var _ = require('underscore'),
     db = require('../db');
 
-var exists = function(val) {
+var empty = function(val) {
```
Comment (alxndrsn) [inline controllers/record-utils.js]: I think this should be called `notEmpty`...

File: controllers/record-utils.js [hunk 2]
```diff
@@ -1,7 +1,7 @@
...
   return val !== '' && typeof val !== 'undefined';
```
Comment (alxndrsn) [inline controllers/record-utils.js]: And should probably handle `null`:
```
> let x = null
> x !== '' && typeof x !== 'undefined'
true
```

File: controllers/record-utils.js [hunk 3]
```diff
@@ -20,11 +20,11 @@ var request = function(opts, callback) {
...
 
 var createByForm = function(data, callback) {
-  var required = ['message', 'from'],
-      optional = ['reported_date', 'locale', 'gateway_ref'];
+  var required = ['from'],
+      optional = ['message', 'reported_date', 'locale', 'gateway_ref'];
   for (var k in required) {
     if (required.hasOwnProperty(k)) {
-      if (!exists(data[required[k]])) {
+      if (!empty(data[required[k]])) {
```
Comment (alxndrsn) [inline controllers/record-utils.js]: ...although here `!notEmpty(…)` is going to be awful to read.

File: controllers/sms-gateway.js [hunk 4]
```diff
@@ -112,6 +112,8 @@ module.exports = {
       getOutgoing
     ], (err, [,,outgoingMessages]) => {
       if (err) {
+        console.error('There was an error processing the following SMS POST request');
+        console.error(JSON.stringify(req.body));
```
Comment (alxndrsn) [inline controllers/sms-gateway.js]: Are you sure you want to log request bodies?  Is there any worry about personally-identifying info with this?

Comment (SCdF) [inline controllers/sms-gateway.js]: I'm definitely not sure I want to do this. It was in fact you who convinced me to do it! I guess what you must have meant was "just for a little bit of time on production" and not in the actual codebase.

File: controllers/record-utils.js (hunk 1 above)
Comment (SCdF) [inline controllers/record-utils.js]: Oh dear.

File: controllers/record-utils.js [hunk 5]
```diff
@@ -20,17 +22,19 @@ var request = function(opts, callback) {
...
-  var required = ['message', 'from'],
-      optional = ['reported_date', 'locale', 'gateway_ref'];
-  for (var k in required) {
-    if (required.hasOwnProperty(k)) {
-      if (!exists(data[required[k]])) {
-        return callback(new Error('Missing required field: ' + required[k]));
-      }
-    }
+  if (empty(data.from)) {
+    return callback(new Error('Missing required field: from'));
```
Comment (alxndrsn) [inline controllers/record-utils.js]: More accurate as `Missing required value: from`?

Comment (SCdF) [inline controllers/record-utils.js]: Hmmm.

Hmmmmmmmmmmmmmmmmmmmmmmmm.

Hmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmm.

Sure.

Comment (alxndrsn) [inline controllers/record-utils.js]: `Missing require value for field from`?

Comment (alxndrsn) [inline controllers/record-utils.js]: etc

Comment (alxndrsn) [inline controllers/record-utils.js]: I was wondering if there was a way to communicate that we needed a value for that field, not just the presence of the field.  Possibly worthwhile as this is the cause of the bug...

**Pull request merged**